### PR TITLE
CAMEL-11333: Create a new camel-thrift RPC component

### PR DESCRIPTION
**Pull request by dmvolod**
Those who will run tests during code review: please pay attention, that the latest libthrift version 0.10.0 using in this component contains regression which is printing debug info to the console out.

### CODE REVIEW

**Comment by davsclaus** [comment]
github says there is conflicting files, wonder if you mind updating the PR

**Comment by dmvolod** [comment]
The conflicts have been fixed, sorry @davsclaus I can't access my local git version to do commits rebase and squash

**Comment by dmvolod** [comment]
@davsclaus , all git issues fixed, please review.

File: components/camel-thrift/src/main/java/org/apache/camel/component/thrift/ThriftConstants.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    String THRIFT_ASYNC_CLIENT_CLASS_NAME = "AsyncClient";
+    String THRIFT_ASYNC_CLIENT_FACTORY_NAME = "Factory";
+    String THRIFT_ASYNC_CLIENT_GETTER_NAME = "getAsyncClient";
+    String THRIFT_SERVER_SYNC_INTERFACE_NAME = "Iface";
+    String THRIFT_SERVER_SYNC_PROCESSOR_CLASS = "Processor";
+    String THRIFT_SERVER_ASYNC_INTERFACE_NAME = "AsyncIface";
+    String THRIFT_SERVER_ASYNC_PROCESSOR_CLASS = "AsyncProcessor";
+    
+    int THRIFT_CONSUMER_POOL_SIZE = 5;
+    int THRIFT_CONSUMER_MAX_POOL_SIZE = 1024;
```
Comment (davsclaus) [inline components/camel-thrift/src/main/java/org/apache/camel/component/thrift/ThriftConstants.java]: If this is threads then 1024 is a very big number isnt it?

File: components/camel-thrift/src/main/java/org/apache/camel/component/thrift/ThriftConsumer.java [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    }
+
+    @Override
+    protected void doStart() throws Exception {
+        super.doStart();
+        if (server == null) {
+            LOG.debug("Starting the Thrift server");
+            initializeServer();
+            server.serve();
+            LOG.info("Thrift server started and listening on port: " + serverTransport.getPort());
```
Comment (davsclaus) [inline components/camel-thrift/src/main/java/org/apache/camel/component/thrift/ThriftConsumer.java]: Use `{}` style for logging

File: components/camel-thrift/src/main/java/org/apache/camel/component/thrift/ThriftProducer.java [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+import org.apache.thrift.transport.TSocket;
+import org.apache.thrift.transport.TTransport;
+import org.apache.thrift.transport.TTransportException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Represents asynchronous and synchronous Thrift producer implementations
+ */
+public class ThriftProducer extends DefaultProducer implements AsyncProcessor {
```
Comment (davsclaus) [inline components/camel-thrift/src/main/java/org/apache/camel/component/thrift/ThriftProducer.java]: You can extend `DefaultAsyncProducer` instead

Comment (davsclaus) [inline components/camel-thrift/src/main/java/org/apache/camel/component/thrift/ThriftProducer.java]: Ah you cannot do that thrift sync vs async api you use a bit further below

File: components/camel-thrift/src/main/java/org/apache/camel/component/thrift/ThriftComponent.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+/**
+ * Represents the component that manages {@link ThriftEndpoint}.
+ */
+public class ThriftComponent extends DefaultComponent {
+
+    protected Endpoint createEndpoint(String uri, String remaining, Map<String, Object> parameters) throws Exception {
+        ThriftConfiguration config = new ThriftConfiguration();
+        setProperties(config, parameters);
+
+        // Extract service and package names from the full service name
```
Comment (dmvolod) [inline components/camel-thrift/src/main/java/org/apache/camel/component/thrift/ThriftComponent.java]: Ok, will fix it.

File: components/camel-thrift/src/main/java/org/apache/camel/component/thrift/ThriftConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+    @UriPath
+    @Metadata(required = "true")
+    private String service;
+    
+    @UriParam(label = "producer")
+    private String method;
+    
+    @UriParam
+    private String host;
```
Comment (dmvolod) [inline components/camel-thrift/src/main/java/org/apache/camel/component/thrift/ThriftConfiguration.java]: Is camel-undertow a good example for this or you could recommend another one?

File: components/camel-thrift/src/main/java/org/apache/camel/component/thrift/ThriftConfiguration.java [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+    @UriParam
+    private int port;
+    
+    @UriParam(label = "consumer", defaultValue = "" + ThriftConstants.THRIFT_CONSUMER_POOL_SIZE)
+    private int poolSize = ThriftConstants.THRIFT_CONSUMER_POOL_SIZE;
+    
+    @UriParam(label = "consumer", defaultValue = "" + ThriftConstants.THRIFT_CONSUMER_MAX_POOL_SIZE)
+    private int maxPoolSize = ThriftConstants.THRIFT_CONSUMER_MAX_POOL_SIZE;
+    
+    private String serviceName;
```
Comment (dmvolod) [inline components/camel-thrift/src/main/java/org/apache/camel/component/thrift/ThriftConfiguration.java]: Ok, will fix it.

File: components/camel-thrift/src/main/java/org/apache/camel/component/thrift/ThriftConstants.java (hunk 1 above)
Comment (dmvolod) [inline components/camel-thrift/src/main/java/org/apache/camel/component/thrift/ThriftConstants.java]: I was wonder too, that the original Thrift server set this value to Integer.MAX_VALUE by default when it creates new ThreadPoolExecutor. Maybe no one used a default values.
What you will recommend for this values?

File: components/camel-thrift/src/main/java/org/apache/camel/component/thrift/ThriftConsumer.java (hunk 2 above)
Comment (dmvolod) [inline components/camel-thrift/src/main/java/org/apache/camel/component/thrift/ThriftConsumer.java]: Ok, will fix it.

File: components/camel-thrift/src/main/java/org/apache/camel/component/thrift/ThriftProducer.java (hunk 3 above)
Comment (dmvolod) [inline components/camel-thrift/src/main/java/org/apache/camel/component/thrift/ThriftProducer.java]: Yes

File: components/camel-thrift/src/main/java/org/apache/camel/component/thrift/ThriftConstants.java (hunk 1 above)
Comment (davsclaus) [inline components/camel-thrift/src/main/java/org/apache/camel/component/thrift/ThriftConstants.java]: Oh is it the size of max threads or max task queue? The latter is the internal task queue where you can store pending tasks to be processed by the thread pool. So you can have eg 10 threads and a task pool of 5000 etc. However a Integer.MAX_VALUE is really too big in all situations. Its better to reject the task which then can work with back pressures in a reactive system

**Comment by dmvolod** [comment]
@davsclaus , I've updated a component, change configuration and uri style to do it more service and cloud friendly.

**Comment by davsclaus** [comment]
Thanks, when I run the unit tests it outputs many 'Received 0' etc. I wonder if you could output this to the logger instead?

**Comment by dmvolod** [comment]
@davsclaus , as I wrote above, this is regression inside thriftlib 0.10.0 which is already fixed, but not released yet.

**Comment by davsclaus** [comment]
Thanks for the PR it has been merged. Do you mind closing this?

**Comment by dmvolod** [comment]
Thanks a lot, @davsclaus , PR closed.

**Pull request closed without merging**